File: Java/300.最长上升子序列.java
import java.util.Arrays;
import java.util.Collections;

/*
 * @lc app=leetcode.cn id=300 lang=java
 *
 * [300] 最长上升子序列
 *
 * 
 * 1.解题思路
 *  DP 
 *  1. subproblem A的最长上升子序列，等于A-1的最长上升子序列+1
 *  2. DP数组 dp[n]
 *  3. DP方程 
 *   if nums[i] > nums[i-1] dp[n] = max(dp[i], dp[j] + 1)
 *   else d[n] = 0
 */


// @lc code=start
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        int[] dp = new int[n+1];
        int maxval = 0;
        dp[0] = 1;
        for (int i=0; i<n; i++) {
            maxval = 0;
            for (int j=0; j<i; j++) {
                if (nums[i] > nums[j]) {
                    maxval = Math.max(maxval, dp[j]); 
                }
            }
            //每轮求当前值的最长上升序列
            dp[i] = maxval + 1;
        }
        
        // 求数组集合里面的最大值
        return Arrays.stream(dp).max().getAsInt();
    }
}
// @lc code=end

File: Python/DP/300.最长上升子序列.py
#-*- coding:utf-8 -*-
# @lc app=leetcode.cn id=   q'q lang=python
#
# [300] 最长上升子序列
#
# 解题思路
# https://leetcode-cn.com/problems/longest-increasing-subsequence/solution/zui-chang-shang-sheng-zi-xu-lie-by-leetcode-soluti/
# dp[i]取的最长升序子序列，必须dp[i-1]为最大，然后加上nums[i]，在满足nums[i]>nums[i-1]的情况下
# 这里有个问题就是不能仅仅和前面的一个元素比较，因为前一个元素可能很小，但是在它这前构建一个上升子串了，
# 前一个元素只是继承，这个时候的比较是错误的，需要从0<=j<i 这样的范围内比较
# 1.dp
#   1.subproblem n的最长上升子序列的长度，等于n-1的最长上升子序列+1
#   2.dp数组 dp[0] = 1 添加第一个为1，就是当前数字也算
#   3.dp方程 
#   if nums[i] > nums[j] dp[i] = max(dp[j]) + 1


# @lc code=start
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if not nums:
            return 0
        
        dp = []
        for i in range(n):
            dp.append(1)
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
    
            
# @lc code=end
if __name__ == "__main__":
    obj = Solution()
    # ret = obj.lengthOfLIS([10,9,2,5,3,4])
    ret = obj.lengthOfLIS([4,10,4,3,8,9])
    print(ret)
    
